api/axis.py

```diff
@@ -448,6 +462,13 @@ def fulfill(self, patron, pin, licensepool, internal_format, **kwargs):
             fulfillment = loan.fulfillment_info
             if not fulfillment or not isinstance(fulfillment, FulfillmentInfo):
                 raise CannotFulfill()
+
+            if isinstance(patron, Patron) and loan.external_identifier:
+                # Save the external_identifier to the Patron's Loan for future retrieval
+                patron_loan = get_one(self._db, Loan, patron=patron, license_pool=licensepool)
+                if patron_loan:
+                    patron_loan.external_identifier = loan.external_identifier
```

Comment:
I was wondering why this code was necessary, because some part of me had assumed this would be taken care of in the API-agnostic code in circulation.py. After looking into it I do think this code is still necessary, but there's a big optimization opportunity in circulation.py, specifically [right here](https://github.com/NYPL-Simplified/circulation/blob/develop/api/circulation.py#L1285).

We can go over this live if you want, since I realize this isn't code we've talked about before, but: `sync_bookshelf` is called periodically by the client refreshing its view of the end-user's loans. It works by calling out to every API that's integrated into the system. Those APIs give it a bunch of `LoanInfo` objects, and it makes sure that the patrons' entries in `loans` are created, deleted, or updated appropriately.

For Axis 360 API, "calling out to the API" means making a call to `availability/v2`. You're changing the `AvailabilityResponseParser` to attach `external_identifier` to every `LoanInfo` that gets parsed.

So, let's say you change `sync_bookshelf` to set `external_identifier` on a `Loan` if there's one in the corresponding `LoanInfo`, similar to how we already set `locked_to` on the `Loan` if the corresponding `LoanInfo` is locked. Now a `sync_bookshelf` call sets `external_identifier` on _every_ active loan that doesn't already have one set. This potentially saves an extra individual `availability/v2` call for each of those loans.

Reply:
Since the known client behavior is to borrow a book and then immediately try to fulfill it (which will set external_identifier for that book), I think this will only improve things in marginal cases. Unless you want to learn more about that part of the code right now, I'm OK with not doing this work in this branch.

Reply:
This would be good to add. I did a quick attempt to see if any of the tests from `test_circulationapi.py` had any loans with an `external_identifier` set, but no loans had one set. I think the work for this could be done in a future branch / ticket.

Reply:
The only other API that _might_ do this is ODL. The `external_identifier` was added to the `loans` table specifically for ODL.